cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp

```diff
@@ -0,0 +1,180 @@
+
+    if (op == SocketOperationNone) // connected
+    {
+        // cancelPrevious = true because Ice often writes to a client connection before it's connected.
+        _timer->schedule(_heartbeatTimerTask, _idleTimeout / 2, true);
+        if (_enableIdleCheck)
+        {
+            // cancelPrevious = true because with SSL, the connection is connected after a read.
+            _timer->schedule(_idleCheckTimerTask, _idleTimeout, true);
+        }
+    }
+
+    return op;
+}
+
+IdleTimeoutTransceiverDecorator::~IdleTimeoutTransceiverDecorator()
+{
+    // If we destroy this object before calling decoratorInit(), _heartbeatTimerTask and _idleCheckTimerTask will be
+    // null.
+    if (_heartbeatTimerTask)
+    {
+        _timer->cancel(_heartbeatTimerTask);
+    }
+    if (_idleCheckTimerTask)
+    {
+        _timer->cancel(_idleCheckTimerTask);
+    }
+}
+
+void
+IdleTimeoutTransceiverDecorator::close()
+{
+    _timer->cancel(_heartbeatTimerTask);
+    if (_enableIdleCheck)
+    {
+        _timer->cancel(_idleCheckTimerTask);
+    }
+    _decoratee->close();
+}
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::write(Buffer& buf)
+{
+    // We're about to write something and we don't want to send a concurrent heartbeat.
+    _timer->cancel(_heartbeatTimerTask);
+
+    SocketOperation op = _decoratee->write(buf);
+    if (op == SocketOperationNone) // write completed
+    {
+        _timer->schedule(_heartbeatTimerTask, _idleTimeout / 2);
+    }
+    return op;
+}
+
+#if defined(ICE_USE_IOCP)
+bool
+IdleTimeoutTransceiverDecorator::startWrite(Buffer& buf)
+{
+    // We're about to write something and we don't want to send a concurrent heartbeat.
+    _timer->cancel(_heartbeatTimerTask);
+
+    return _decoratee->startWrite(buf);
+
+    // TODO: should we schedule the next heartbeat when the return value is true? Or do we always call finishWrite?
+}
+
+void
+IdleTimeoutTransceiverDecorator::finishWrite(Buffer& buf)
+{
+    _decoratee->finishWrite(buf);
+
+    // Schedule a heartbeat after writing all the data.
+    if (buf.i == buf.b.end())
+    {
+        _timer->schedule(_heartbeatTimerTask, _idleTimeout / 2);
+    }
+}
+
+void
+IdleTimeoutTransceiverDecorator::startRead(Buffer& buf)
+{
+    // We always call finishRead or read to actually read the data.
+    _decoratee->startRead(buf);
+}
+
+void
+IdleTimeoutTransceiverDecorator::finishRead(Buffer& buf)
+{
+    if (_enableIdleCheck)
+    {
+        // We don't want the idle check to run while we're reading, so we reschedule it before reading.
+        _timer->schedule(_idleCheckTimerTask, _idleTimeout, true);
+    }
+
+    _decoratee->finishRead(buf);
+}
+
+#endif
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::read(Buffer& buf)
+{
+    if (_enableIdleCheck)
+    {
+        // We don't want the idle check to run while we're reading, so we reschedule it before reading.
+        _timer->schedule(_idleCheckTimerTask, _idleTimeout, true);
+    }
+    return _decoratee->read(buf);
```

Comment:
The read and write logic are not the same:

- for write, we don't want to write more (send a heartbeat) if another write is already in progress, even if this write is very slow / stuck. It would be pointless. That's why we schedule a heartbeat only once the write is completed.
- for read, we want to run the idle check even while a read is in progress, if this read is stuck

So I think the read logic is correct. We get here (in read) only when there is something new to read.

Reply:
> for read, we want to run the idle check even while a read is in progress, if this read is stuck

All the transceiver methods (connect, initialize, read, write) should be block. It's the contract.

Reply:
By "stuck" I meant: the next read call is not called in a timely manner. Each read call always completes quickly.